### PR TITLE
Date detection should not rely on a hardcoded set of characters.

### LINKED ISSUE #1694: Add `T` (on top of `:`, `-`, and `/`) as a way to pre detect of a string is a date for dynamic date formats

### PR DESCRIPTION
**Pull request by jpountz**
Currently we only apply date detection on strings that contain either `:`, `-`
or `/`. This commit inverses the heuristic in order to only apply date detection
on strings that are not parseable as a number, so that more date formats can be
used as dynamic dates formats.

Closes #1694

### CODE REVIEW

**Comment by jpountz** [comment]
@elasticmachine retest this please

File: core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java [hunk 1]
```diff
@@ -650,47 +649,55 @@ private static void parseNullValue(ParseContext context, ObjectMapper parentMapp
...
+            } else if (parseableAsDouble && context.root().numericDetection()) {
+                Mapper.Builder builder = context.root().findTemplateBuilder(context, currentFieldName, XContentFieldType.DOUBLE);
+                if (builder == null) {
+                    builder = newFloatBuilder(currentFieldName, Version.indexCreated(context.indexSettings()));
+                }
+                return builder;
+            } else if (parseableAsLong == false && parseableAsDouble == false && context.root().dateDetection()) {
+                // We refuse to match pure numbers, which are too likely to be
+                // false positives with date formats that include eg.
+                // `epoch_millis` or `YYYY`
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java]: I guess it's difficult to detect ? If the pattern is `epoch_millis` or composed of number symbol only ? If it's not detectable then maybe add a small warning in the docs.

**Review by jimczi — COMMENTED** [review]
The change looks good.
To make the thing clear you could maybe add a test with a pattern that was not detected before and one with a pattern composed of numbers only that should be detected as a number.

File: core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java (hunk 1 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java]: Yes, I don't think we can detect this confidently. I'll work on your test suggestions.

**Comment by jpountz** [comment]
@jimczi Can you have another look?

**Review by jimczi — APPROVED** [review]
Thanks adding the tests.
LGTM

**Comment by jpountz** [comment]
NOTE: I'm thinking of only having this change in 6.0 as this might be something surprising to get in a minor release for some of our users. Please speak up if you think I should merge to 5.2.

**Comment by jimczi** [comment]
I think it should be merged to 5.2. I don't think that people rely on the fact that some pattern are not detected with dates. Although it's for dynamic detection so this won't change the old indices but could be beneficial for new indices ?

**Comment by jpountz** [comment]
It could change old indices as well if a new field is introduced. But indeed, I would expect most old indices to already have a "stable" mapping. @clintongormley I'd like to run it by you before making a decision, what do you think?

**Comment by clintongormley** [comment]
It has been like this since the first version, I think it is ok to wait a bit longer for 6.0

**Pull request merged**